handlers/handlers.go

```diff
@@ -89,14 +111,92 @@ func HandleSessionRequest(
 			panic(err)
 		}
 
+		// Create token as a cookie and set it
+		{
+			userId, err := userservice.GetUserIdFromEmailAddress(loginForm.EmailAddress)
+			if err != nil {
+				panic(err)
+			}
+
+			token, err := createTokenAsString(userId, oneWeekInMinutes)
+			if err != nil {
+				panic(err)
+			}
+
+			expiration := time.Now().Add(oneWeekInMinutes * time.Minute)
+
+			cookie := http.Cookie{
+				Name:     cerealNotesCookieName,
+				Value:    token,
+				Expires:  expiration,
+				HttpOnly: true,
+			}
+
+			http.SetCookie(responseWriter, &cookie)
+
+		}
+
 		responseWriter.WriteHeader(http.StatusCreated)
 		responseWriter.Write([]byte(fmt.Sprint("passward email combo was correct")))
 
+	case http.MethodDelete:
+		//Cookie will overwrite existing cookie then delete itself
+		cookie := http.Cookie{
+				Name:     cerealNotesCookieName,
+				Value:    "",
+				HttpOnly: true,
+				MaxAge:   -1,
+			}
+
+		http.SetCookie(responseWriter, &cookie)
+		responseWriter.WriteHeader(http.StatusOK)
+		responseWriter.Write([]byte(fmt.Sprint("user succefully logged out")))
 	default:
 		respondWithMethodNotAllowed(responseWriter, []string{http.MethodPost})
 	}
 }
 
+type AuthentictedRequestHandlerType func(http.ResponseWriter, *http.Request, models.UserId)
+
+func AuthenticateOrRedirectToLogin(
+	authenticatedHandlerFunc AuthentictedRequestHandlerType,
+) func(http.ResponseWriter, *http.Request) {
+	return http.HandlerFunc(
```

Comment:
I think this is actually a golang cast this anonymous function into a HandlerFunc type.

Reply:
why is that necessary though? The function you're passing it has the same signature doesn't it?

Reply:
I tried compiling it without the cast and it works. I got this code from a website and I think this is in case a function caller requires a http.HanderFunc type. Since go type checks even if they are the same type, it won't implicitly "upcast" a type. So for example if the function demanded an input as UserIdType and you gave it an int64 it would throw an error unless you cast it, however, if the function required an int64 and you gave it a UserId i don't think it would throw an error. I mild preference for leaving it, but can be easily convinced to remove it.